File: src/ambilight/gui/HideableFrame.java
package ambilight.gui;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class HideableFrame extends JFrame {

	protected boolean minimizeAutomatically;
	private TrayIcon trayIcon;
	private Point mouseClickPoint;

	public HideableFrame(String title, String iconPath) {
		super(title);

		Image icon = new ImageIcon(this.getClass().getResource(iconPath)).getImage();
		setIconImage(icon);

		if (SystemTray.isSupported()) {

			MenuItem openItem = new MenuItem("Open");
			openItem.addActionListener(e -> {
				showFrame();
			});

			MenuItem exitItem = new MenuItem("Exit");
			exitItem.addActionListener(e -> {
				SystemTray.getSystemTray().remove(trayIcon);
				System.exit(0);
			});

			PopupMenu popup = new PopupMenu();
			popup.add(openItem);
			popup.add(exitItem);

			trayIcon = new TrayIcon(icon, title, popup);
			trayIcon.setImageAutoSize(true);
			trayIcon.addMouseListener(new MouseAdapter() {
				@Override
				public void mouseClicked(MouseEvent e) {
					super.mouseClicked(e);
					if (e.getButton() == MouseEvent.BUTTON1) {
						if (isShowing()) {
							hideFrame();
						} else {
							showFrame();
						}
					}
				}
			});


			try {
				SystemTray.getSystemTray().add(trayIcon);
			} catch (AWTException e) {
				e.printStackTrace();
			}

			addWindowStateListener(new WindowStateListener() {
				@Override
				public void windowStateChanged(WindowEvent e) {
					if ((e.getNewState() & ICONIFIED) == ICONIFIED) {
						setVisible(false);
					} else {
						setVisible(true);
					}
				}
			});

			minimizeAutomatically = true;
			addWindowFocusListener(new WindowFocusListener() {
				@Override
				public void windowGainedFocus(WindowEvent e) {
					if (minimizeAutomatically)
						showFrame();
				}

				@Override
				public void windowLostFocus(WindowEvent e) {
					if (minimizeAutomatically)
						hideFrame();
				}
			});

			setDefaultCloseOperation(WindowConstants.HIDE_ON_CLOSE);

		} else {
			System.out.println("system tray not supported");
			setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
		}

		addMouseListener(new MouseAdapter() {
			public void mousePressed(MouseEvent e) {
				mouseClickPoint = e.getPoint();
			}
		});

		addMouseMotionListener(new MouseAdapter() {
			public void mouseDragged(MouseEvent e) {
				int X = getLocation().x + e.getX() - mouseClickPoint.x;
				int Y = getLocation().y + e.getY() - mouseClickPoint.y;
				setLocation(X, Y);
			}
		});

	}

	public void resetLocation() {
		setLocation(Toolkit.getDefaultToolkit().getScreenSize().width - getWidth() - 16,
					Toolkit.getDefaultToolkit().getScreenSize().height - getHeight() - 56);
	}

	public void setMinimizeAutomatically(boolean minimizeAutomatically) {
		this.minimizeAutomatically = minimizeAutomatically;
	}

	protected void showFrame() {
//		System.out.println("Show");
		setExtendedState(JFrame.NORMAL);
		setVisible(true);
		toFront();
//		requestFocus();
	}

	protected void hideFrame() {
//		System.out.println("Hide");
		setExtendedState(JFrame.ICONIFIED);
		setVisible(false);
	}

	protected void exit() {
		if (trayIcon != null)
			SystemTray.getSystemTray().remove(trayIcon);
		System.exit(0);
	}

}

File: gui/src/main/java/me/khol/ambilight/gui/HideableFrame.kt
package me.khol.ambilight.gui

import java.awt.*
import java.awt.event.*
import javax.swing.*

open class HideableFrame(title: String, iconPath: String) : JFrame(title) {

	protected var minimizeAutomatically: Boolean = false
	private lateinit var trayIcon: TrayIcon
	private lateinit var mouseClickPoint: Point

	init {
		val icon = ImageIcon(this.javaClass.getResource(iconPath)).image
		iconImage = icon

		if (SystemTray.isSupported()) {

			val popup = PopupMenu().apply {
				add(MenuItem("Open").apply {
					addActionListener {
						showFrame()
					}
				})
				add(MenuItem("Exit").apply {
					addActionListener {
						SystemTray.getSystemTray().remove(trayIcon)
						System.exit(0)
					}
				})
			}

			trayIcon = TrayIcon(icon, title, popup).apply {
				isImageAutoSize = true
				addMouseListener(object : MouseAdapter() {
					override fun mouseClicked(e: MouseEvent) {
						super.mouseClicked(e)
						if (e.button == MouseEvent.BUTTON1) {
							if (isShowing) {
								hideFrame()
							} else {
								showFrame()
							}
						}
					}
				})
			}

			try {
				SystemTray.getSystemTray().add(trayIcon)
			} catch (e: AWTException) {
				e.printStackTrace()
			}

			addWindowStateListener { e ->
				isVisible = e.newState and Frame.ICONIFIED != Frame.ICONIFIED
			}

			minimizeAutomatically = true
			addWindowFocusListener(object : WindowFocusListener {
				override fun windowGainedFocus(e: WindowEvent) {
					if (minimizeAutomatically) {
						showFrame()
					}
				}

				override fun windowLostFocus(e: WindowEvent) {
					if (minimizeAutomatically) {
						hideFrame()
					}
				}
			})

			defaultCloseOperation = WindowConstants.HIDE_ON_CLOSE

		} else {
			println("system tray not supported")
			defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
		}

		addMouseListener(object : MouseAdapter() {
			override fun mousePressed(e: MouseEvent) {
				mouseClickPoint = e.point
			}
		})

		addMouseMotionListener(object : MouseAdapter() {
			override fun mouseDragged(e: MouseEvent) {
				val x = location.x + e.x - mouseClickPoint.x
				val y = location.y + e.y - mouseClickPoint.y
				setLocation(x, y)
			}
		})
	}

	fun resetLocation() {
		setLocation(
			Toolkit.getDefaultToolkit().screenSize.width - width - 16,
			Toolkit.getDefaultToolkit().screenSize.height - height - 56
		)
	}


	protected fun showFrame() {
		extendedState = JFrame.NORMAL
		isVisible = true
		toFront()
//		requestFocus();
	}

	protected fun hideFrame() {
		extendedState = JFrame.ICONIFIED
		isVisible = false
	}

	protected fun exit() {
		if (SystemTray.isSupported()) {
			SystemTray.getSystemTray().remove(trayIcon)
		}
		System.exit(0)
	}
}